opencog/cogserver/server/AgentRunnerThread.cc

```diff
@@ -0,0 +1,167 @@
+/*
+ * opencog/server/AgentRunnerThread.cc
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License v3 as
+ * published by the Free Software Foundation and including the exceptions
+ * at http://opencog.org/wiki/Licenses
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program; if not, write to:
+ * Free Software Foundation, Inc.,
+ * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ */
+
+#include <opencog/cogserver/server/AgentRunnerThread.h>
+#include <opencog/util/Logger.h>
+
+using namespace std;
+
+
+namespace opencog
+{
+
+AgentRunnerThread::AgentRunnerThread(const std::string &name) :
+        AgentRunnerBase(name), runAgents(false), agentsModified(false),
+        clearAll(false)
+{
+}
+
+AgentRunnerThread::~AgentRunnerThread()
+{
+    removeAllAgents();
+    joinRunThread();
+}
+
+void AgentRunnerThread::start()
+{
+    {
+        // locking is required, to prevent dead-lock in condition_variable wait()
+        lock_guard<mutex> lock(runAgentsMutex);
+        runAgents.store(true, memory_order_relaxed);
+    }
+    runAgentsCond.notify_one();
+}
+
+void AgentRunnerThread::stop()
+{
+    runAgents.store(false, memory_order_relaxed);
+}
+
+void AgentRunnerThread::addAgent(AgentPtr a)
+{
+    if (!hasAgents()) {
+        joinRunThread();
+        /*
+         * processAgentsThread() will exit immediately if no agents are available,
+         * so we add the first one immediately
+         */
+        {
+            lock_guard<mutex> lock(agentsMutex);
```

Comment:
Why is AgentRunnerBase not thread safe? Don't you think it would be wise to define it so that its thread safe?

Neither addAgent() nor push_back() are resources. An std::vector is a resource. Please review the concept of a resource, here:   https://en.wikipedia.org/wiki/System_resource

and how resources can be protected with locks: https://en.wikipedia.org/wiki/Lock_(computer_science)

If you are not thinking about resources and locks, you are thinking about the problem incorrectly.